engine/apps/alerts/tasks/notify_user.py

```diff
@@ -418,17 +487,119 @@ def perform_notification(log_record_pk):
 
 
 @shared_dedicated_queue_retry_task(
-    autoretry_for=(Exception,), retry_backoff=True, max_retries=0 if settings.DEBUG else None
+    autoretry_for=(Exception,), retry_backoff=True, max_retries=1 if settings.DEBUG else None
 )
-def send_user_notification_signal(log_record_pk):
-    start_time = time.time()
+def send_bundled_notification(user_notification_bundle_id):
+    from apps.alerts.models import AlertGroup, UserNotificationBundle
+    from apps.base.models import UserNotificationPolicy, UserNotificationPolicyLogRecord
+
+    task_logger.info(f"Start send_bundled_notification for user_notification_bundle {user_notification_bundle_id}")
+    with transaction.atomic():
+        user_notification_bundle = UserNotificationBundle.objects.filter(
+            pk=user_notification_bundle_id
+        ).select_for_update()[0]
+
+        if send_bundled_notification.request.id != user_notification_bundle.notification_task_id:
+            task_logger.info(
+                f"send_bundled_notification: notification_task_id mismatch. "
+                f"Duplication or non-active notification triggered. "
+                f"Active: {user_notification_bundle.notification_task_id}"
+            )
+            return
+
+        alert_group_ids = [i["alert_group_id"] for i in user_notification_bundle.notification_data]
+        notification_policy_ids = [i["notification_policy_id"] for i in user_notification_bundle.notification_data]
+        # get active alert groups to notify about
+        active_alert_groups = set(
+            AlertGroup.objects.filter(
+                id__in=alert_group_ids,
+                resolved=False,
+                acknowledged=False,
+                silenced=False,
+            ).values_list("id", flat=True)
+        )
+        existing_notification_policies = set(
+            UserNotificationPolicy.objects.filter(id__in=notification_policy_ids).values_list("id", flat=True)
+        )
+        current_notification_data = []
+        created_lod_records = []
+
+        for entry in user_notification_bundle.notification_data:
+            if entry["alert_group_id"] not in active_alert_groups:
+                task_logger.info(
+                    f"alert_group {entry['alert_group_id']} is not active or doesn't exist, skip notification"
```

Comment:
👍 nice! this will come in handy for those users that are deleting groups in the API before we handle them.